src/station_gui.cpp

```diff
@@ -655,11 +606,45 @@ class CompanyStationsWindow : public Window
 				this->stations.SetSortType(index);
 
 				/* Display the current sort variant */
-				this->GetWidget<NWidgetCore>(WID_STL_SORTDROPBTN)->widget_data = CompanyStationsWindow::sorter_names[this->stations.SortType()];
+				this->GetWidget<NWidgetCore>(WID_STL_SORTDROPBTN)->widget_data = this->sorter_names[this->stations.SortType()];
 
 				this->SetDirty();
 			}
 		}
+
+		if (widget == WID_STL_CARGODROPDOWN) {
+			FilterState oldstate = this->filter;
+
+			if (index >= 0 && index < NUM_CARGO) {
+				if (_ctrl_pressed) {
+					ToggleBit(this->filter.cargoes, index);
+				} else {
+					this->filter.cargoes = 1ULL << index;
+					this->filter.include_empty = false;
+				}
+			} else if (index == NUM_CARGO) {
```

Comment:
I would alias the `NUM_CARGO` where `CF_SELECT_ALL` is created, to make a distinction between the actual number of cargoes or the "include empty" being selected or not. The `NUM_CARGO` for the previous state is fine.